test/distributed/pipelining/test_schedule_multiproc.py

```diff
@@ -932,6 +934,201 @@ def dw_runner():
+    def test_schedule_memory_usage_zero_bubble(self, ScheduleClass):
+        stages_per_rank = 2
+        n_stages = stages_per_rank * self.world_size
+
+        mod = MultiMLP(d_hid, n_layers=n_stages)
+        mod.to(self.device)
+
+        x = torch.randn(batch_size, d_hid, device=self.device)
+
+        # pre-run to allocate cuBLAS workspaces (see in test_fully_shard_memory.py)
+        mod(x).sum().backward()
+        torch.cuda.empty_cache()
+
+        x = torch.randn(batch_size, d_hid, device=self.device)
+        target = torch.randn(batch_size, d_hid, device=self.device)
+
+        mod.zero_grad(set_to_none=True)  # reset grads memory
+        base_mem = self._get_curr_active_memory()
+
+        loss_fn = torch.nn.MSELoss(reduction="sum")
+
+        chunks = 4
+
+        stage_indices = [
+            self.rank + i * self.world_size for i in range(stages_per_rank)
+        ]
+        submod_names = [f"layers.{i}" for i in stage_indices]
+        stage_modules = [mod.get_submodule(submod_name) for submod_name in submod_names]
+
+        my_params_size = (
+            sum(
+                sum(p.numel() * p.element_size() for p in part.parameters())
+                for part in stage_modules
+            )
+            / 1e6
+        )
+
+        stages = [
+            PipelineStage(
+                stage_module,
+                stage_idx,
+                n_stages,
+                self.device,
+            )
+            for stage_module, stage_idx in zip(stage_modules, stage_indices)
+        ]
+
+        schedule = ScheduleClass(stages, chunks, loss_fn=loss_fn)
+
+        current_mem = self._get_curr_active_memory()
+        self.assertLessEqual(
+            current_mem,
+            base_mem,
+            f"Rank {self.rank}: Creating the pipeline should not increase memory usage",
+        )
+
+        buffer_mem = x.numel() * x.element_size() / 1e6
+
+        if self.rank == 0:
+            schedule.step(x)
+            output_mem = 0
+
+            buffer_mem *= 3  # recv grads (0) + recv acts (2) + recv grads (2)
+        elif self.rank == self.world_size - 1:
+            losses = []
+            output = schedule.step(target=target, losses=losses)
+
+            output_mem = (output.numel() * output.element_size()) / 1e6
+            buffer_mem *= 3  # recv acts (1) + recv acts (3) + recv grads (1)
+        else:
+            schedule.step()
+            output_mem = 0
+
+        # these states are cleared at the beginning of each step, but not the end, so we need to clear them here
+        for stage in stages:
+            stage.clear_runtime_states()
+
+        gc.collect()
+        torch.cuda.empty_cache()
+
+        current_mem = self._get_curr_active_memory()
+
+        expected_mem = math.ceil(base_mem + output_mem + my_params_size + buffer_mem)
+        print(
+            f"Rank {self.rank} current_mem: {current_mem} ; expected: {base_mem} + {output_mem} + {my_params_size} + {buffer_mem} = {expected_mem}"
```

Comment:
This should fix the CI lint error:

`pip install lintrunner`
`lintrunner -a`

(https://github.com/suo/lintrunner)